docs/design/uplink-scopes.md

```diff
@@ -8,20 +8,17 @@ This design document proposes the data we need to store inside of a share in ord
 
 In order to access objects, one must provide an API key to a given satellite. This API key is tied to a project which allows the satellite to look up the specific bucket being requested and do the things it needs to do. API keys are moving to be macaroons where they can be limited to only allow reads or writes, or expire, etc. Those features can easily be implemented and don't interact much the overall uplink command experience: you have access to do some actions or you don't.
 
-There are complications when it comes to restricting to a whitelist of buckets and/or path prefixes inside of those buckets. For example, if a macaroon is limited to only allow access to paths beneath `sj://bucket/a/b`, then it would be impossible to list the root. Additionally, the clients and servers all operate on _encrypted_ paths, making it impossible for the server to issue the list and filter, because one would have to reveal the root encryption key for the bucket for the client to be able to decrypt the names.
-
-One could attempt to cause restricted macaroons to be "mounted". For example, if a bucket contains `sj://bucket/a/b/c` then the listing of the root would return "c". Unfortunately, this does not compose well with having multiple allowed paths in a macaroon: which "root" do you mean? It also causes confusion because the paths of objects inside of a bucket becomes a local rather than global property, even though the clients and servers must speak with the global view at all times.
-
-Additionally, there's a single global key per project. That's less than ideal for a situation like having distinct prod and staging buckets: you don't want them both using the same encryption keys so that giving access to staging does not leak the encryption keys for prod. It also doesn't allow one to re-key a bucket if a share is leaked. Even if the macaroon inside is revoked, it may be seen as a risk that the key was leaked because they are now depending on the satellite operator for the safety of their data.
+There are complications when it comes to restricting to a whitelist of buckets and/or path prefixes inside of those buckets. For example, if a macaroon is limited to only allow access to paths beneath `sj://bucket/a/b`, then it would be impossible to list the root. Additionally, the clients and servers all operate on **encrypted** paths, making it impossible for the server to issue the list and filter, because one would have to reveal the root encryption key for the bucket for the client to be able to decrypt the names.
 
 ## Design
 
-Uplink will operate on a set of "scopes". Each scope contains an API key, a satellite url, and a list of bucket shares. In Go,
+Uplink will operate on a set of "scopes". Each scope contains an API key, a satellite url, an optional project key, and a list of bucket shares. In Go,
```

Comment:
I'm concluding that this project key is the project-level encryption key, not an API-type key, if it is used to derive bucket keys (which are again, presumably, encryption keys).

I'd assumed we would handle sharing at the project level by having a `BucketShare` with the empty string (or maybe `"/"`?) for the `Bucket` and `EncryptedPath` and `UnencryptedPath` fields. But your way seems fine too. I only wonder whether we shouldn't have `ProjectKey` be a `[]byte` instead of `string`, since `EncryptionKey` is `[]byte` in `BucketShare`.

Reply:
Hmm... We could encode it by having the empty bucket be for the project-level encryption key, but then yeah the path fields are kinda meaningless. It's a special case either way I think, so I prefer the one that's more obvious to a reader, which IMO is an explicit field with documentation.

You are correct that it should be a `[]byte`.